File: apiv2/web/models/tile.ex
defmodule Apiv2.Tile do
  use Apiv2.Web, :model

  schema "apiv1_tiles" do
    field :tile_type, :string
    field :tile_name, :string
    field :status, :string
    field :x, :integer
    field :y, :integer
    field :z, :integer
    field :width, :decimal
    field :height, :decimal
    field :deleted_at, Ecto.DateTime
    field :fire_id, :string

    has_many :cameras, Apiv2.Camera, foreign_key: :tile_id
    timestamps inserted_at: :created_at
  end

  @required_fields ~w(tile_type x y)
  @optional_fields ~w(deleted_at tile_name fire_id status z width height)

  @doc """
  Creates a changeset based on the `model` and `params`.

  If `params` are nil, an invalid changeset is returned
  with no validation performed.
  """
  def changeset(model, params \\ :empty) do
    model
    |> cast(params, @required_fields, @optional_fields)
  end
end

File: apiv1/app/models/apiv1/tile.rb
# == Schema Information
#
# Table name: apiv1_tiles
#
#  id         :integer          not null, primary key
#  tile_type  :string(255)
#  tile_name  :string(255)
#  status     :string(255)
#  x          :integer
#  y          :integer
#  z          :integer
#  created_at :datetime         not null
#  updated_at :datetime         not null
#  deleted_at :datetime
#  fire_id    :string(255)
#  width      :decimal(10, 6)
#  height     :decimal(10, 6)
#

class Apiv1::Tile < ActiveRecord::Base
  KnownTypes = [:barn, :scale, :warehouse, :road, :station]
  acts_as_paranoid
  validates :tile_type,
    presence: true,
    inclusion: { in: KnownTypes.map(&:to_s) }

  scope :living_barns,
    -> { where tile_type: :barn }
  scope :living_scales,
    -> { where tile_type: :scale }
  scope :living_warehouses,
    -> { where tile_type: :warehouse }
  scope :living_roads,
    -> { where tile_type: :road }
  scope :living_stations,
    -> { where tile_type: :station }

  has_many :cameras,
    class_name: "Apiv1::Camera",
    dependent: :destroy

  def master_attributes
    {
      id: id,
      rail_id: id,
      fire_id: fire_id,
      x: x,
      y: y
    }
  end

  def ember_attributes
    attributes.to_h.tap do |a|
      a[:width] ||= 1
      a[:height] ||= 1
      a[:cameras] ||= cameras.map(&:id)
    end
  end

  def ember_json
    {
      tile: ember_attributes,
      cameras: cameras.map(&:ember_attributes)
    }
  end
end
